include/boost/math/differentiation/autodiff.hpp

```diff
@@ -1291,12 +1288,13 @@ fvar<RealType,Order> sqrt(const fvar<RealType,Order>& cr)
   {
     root_type numerator = 0.5;
     root_type powers = 1;
-    derivatives[1] = numerator / *derivatives;
-    for (size_t i=2 ; i<=order ; ++i)
+    for (size_t i=1 ; i<=order ; ++i)
     {
-      numerator *= -0.5 * ((i<<1)-3);
-      powers *= x;
       derivatives[i] = numerator / (powers * *derivatives);
+      if (i < order) {
```

Comment:
This is another downgrade for C++17.

I guess I'm not getting my point across about not making these kinds of compromises. I'll try a better explanation later.

Reply:
Are you referring to `if (i < order)` in particular? If so, I'm not sure what can be done in C++17 as opposed to earlier standards for this line.

Reply:
I think the point is that here the algorithm has been complicated for all compilers in order to compensate for an issue in non-c++17 compilers. If your primary goal is good code for c++17 and your secondary goal is support for c++11 then this is not acceptable. I wonder though if there is a way to make this work in c++11 without at the very least making the code more complicated for the c++17 reader. i.e. if someone who is compiling in c++17 is reading this code then any additional complication to the code would be worsening that c++17 user's experience.

As for this particular change and independent from pulver's point, are you confident that this condition is actually cheaper than the redundant work it protects against? The most expensive thing here is a couple of floating point multiplications. I was under the impression that a jump would be comparatively expensive. At the very least it's a jump which wasn't present in the original version. Full disclosure: I've made no attempt to profile any of the approaches.

Reply:
The alternative to making all non-zero array indices non-literal for non-C++17 is to separate the clauses into separate structs (and using tag dispatching or SFINAE), which is perfectly reasonable. I chose this way to deal with the issue you found since it changed the original code the least (by moving the write to index 1 into the loop and updating the two variables within the loop after assignment).

As for the expense of the jump, I was considering a worst than arithmetic root type (ie boost::multiprecision::cpp_bin_float_50), for which the extra multiplication/assignment operations would be less trivial. GCC offers functionality to help with these jumps, but I’m not aware of MSVC/Clang supporting them.

Reply:
I would much prefer this change:
```
#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
    derivatives[1] = numerator / *derivatives;
#else // for compilers that compile this branch when order=0.
    derivatives[std::min(size_t(1),order)] = numerator / *derivatives;
#endif
```
I don't wish to ever sacrifice (1) performance or (2) code expression/brevity offered by C++17 for the sake of backward compatibility.

Using `#ifndef BOOST_NO_CXX17_IF_CONSTEXPR` may seem ugly, and perhaps seemingly contradictory to (2), but it is a way to preserve the ideal C++17 version so that when the time comes to drop C++11 (and eventually C++14) support, these #conditionals wil make it much easier. It's also a good way to demonstrate the improvement that C++17 has over previous versions.

In this particular case, there is a slight performance degradation in checking the `i < order` conditional, and extra increment `i+1`, in every iteration which is not in the original version.

Reply:
np, I'll update the PR

Reply:
On the subject of the apparent contradiction to (2) have you considered simply ifdefing the entire method and putting the c++11 version in a separate header? That way if you're working in c++17 you can simply ignore that header entirely and all you have to complicate things is a conditional that's always true. I notice you already have a c++11 header. I'm not sure if this is its purpose. Generally speaking it would be bad practice to duplicate an entire method just to tweak a small portion of it but in this case it might be the best option in terms of providing the optimal c++17 experience.